### PR TITLE
Add IncludeTotalCount overloading method

### PR DESCRIPTION
**Pull request by KenitoInc**
### Issues

*This pull request fixes issue #1334 .*

### Description
By default, IncludeTotalCount() method handles $count=true only.
We have added a new method IncludeTotalCount(bool countQuery) that accepts a bool parameter. This parameter handles $count=true and $count=false in query options.

*Support for OData v4*
In OData v4, we return `Count` instead of `TotalCount`, we have added functionality to get `Count` from `QueryOperationResponse<T>`. We have deprecated `TotalCount`
We have added methods IncludeCount() and IncludeCount(bool countQuery) and deprecated `IncludeTotalCount`

So the customer should execute the code as follows:
```
DataServiceQuery<Person> query = dsc.CreateQuery<Person>("People");
query = query.IncludeCount(true);
QueryOperationResponse<Person> response = await query.ExecuteAsync() as QueryOperationResponse<Person>;
Console.WriteLine($"There are a total of {response.Count} People.");
```

### Checklist (Uncheck if it is not completed)

- [x] *Test cases added*
- [ ] *Build and test with one-click build and test script passed*

### Additional work necessary

*If documentation update is needed, please add "Docs Needed" label to the issue and provide details about the required document change in the issue.*

### CODE REVIEW

File: src/Microsoft.OData.Client/DataServiceQueryOfT.cs [hunk 1]
```diff
@@ -314,14 +314,27 @@ public DataServiceQuery<TElement> Expand<TTarget>(Expression<Func<TElement, TTar
...
             return (DataServiceQuery<TElement>)this.Provider.CreateQuery<TElement>(
                 Expression.Call(
                     Expression.Convert(this.Expression, typeof(DataServiceQuery<TElement>.DataServiceOrderedQuery)),
                     mi));
         }
 
+        /// <summary>Requests that the count of all entities in the entity set be returned inline with the query results.</summary>
+        /// <returns>A new <see cref="T:Microsoft.OData.Client.DataServiceQuery`1" /> object that has the inline count option set.</returns>
+        /// <param name="includeTotal">Whether to include total count.</param>
+        public DataServiceQuery<TElement> IncludeTotalCount(bool includeTotal)
```
Comment (g2mula) [inline src/Microsoft.OData.Client/DataServiceQueryOfT.cs]: The 
C# provides an optional arguments feature which may have some candidate use here
The signature could be
```public DataServiceQuery<TElement> IncludeTotalCount(bool? includeTotal = false)```
[More info](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/named-and-optional-arguments#optional-arguments)
However considering that you are doing some reflection stuff to attach the parameterless method above to the expression, maybe old school overloading is how to do it...

File: src/Microsoft.OData.Client/DataServiceQueryOfT.cs [hunk 2]
```diff
@@ -314,14 +314,27 @@ public DataServiceQuery<TElement> Expand<TTarget>(Expression<Func<TElement, TTar
         /// <returns>A new <see cref="T:Microsoft.OData.Client.DataServiceQuery`1" /> object that has the inline count option set.</returns>
         public DataServiceQuery<TElement> IncludeTotalCount()
         {
-            MethodInfo mi = typeof(DataServiceQuery<TElement>).GetMethod("IncludeTotalCount");
+            MethodInfo mi = typeof(DataServiceQuery<TElement>).GetMethods()
+                .FirstOrDefault(m => m.Name == "IncludeTotalCount" && m.GetParameters().Length == 0);
```
Comment (g2mula) [inline src/Microsoft.OData.Client/DataServiceQueryOfT.cs]: Consider using ```nameof(IncludeTotalCount)``` as opposed to the string ```"IncludeTotalCount"``` if the name were ever to change, a refactor will be caught and fixed.
[More info](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/nameof)

Comment (g2mula) [inline src/Microsoft.OData.Client/DataServiceQueryOfT.cs]: After opting for ```.FirstOrDefault``` you would need to take care of the ```OrDefault``` part of the result, what if ```mi == null```?

Comment (xuzhg) [reply src/Microsoft.OData.Client/DataServiceQueryOfT.cs]: +1

---
In reply to: [375320794](https://github.com/OData/odata.net/pull/1658#discussion_r375320794) [](ancestors = 375320794)

File: src/Microsoft.OData.Client/DataServiceQueryOfT.cs [hunk 3]
```diff
@@ -314,14 +314,27 @@ public DataServiceQuery<TElement> Expand<TTarget>(Expression<Func<TElement, TTar
...
         public DataServiceQuery<TElement> IncludeTotalCount()
         {
-            MethodInfo mi = typeof(DataServiceQuery<TElement>).GetMethod("IncludeTotalCount");
+            MethodInfo mi = typeof(DataServiceQuery<TElement>).GetMethods()
+                .FirstOrDefault(m => m.Name == "IncludeTotalCount" && m.GetParameters().Length == 0);
 
             return (DataServiceQuery<TElement>)this.Provider.CreateQuery<TElement>(
                 Expression.Call(
                     Expression.Convert(this.Expression, typeof(DataServiceQuery<TElement>.DataServiceOrderedQuery)),
                     mi));
```
Comment (xuzhg) [inline src/Microsoft.OData.Client/DataServiceQueryOfT.cs]: It's better to move all the code into the new overload method. In this old method, call the new one with the default "bool" value.

File: test/EndToEndTests/Tests/Client/Build.Desktop/QueryCountTests.cs [hunk 4]
```diff
@@ -56,6 +56,21 @@ public void CountLinqTest()
...
 
+        /// <summary>
+        /// IncludeTotalCount(true) Test
+        /// </summary>
+        [TestMethod]
+        public void CountWithBoolTrueParamLinqTest()
+        {
+            var context = this.CreateWrappedContext<DefaultContainer>().Context;
+
+            var query = context.Customer.IncludeTotalCount(true);
```
Comment (xuzhg) [inline test/EndToEndTests/Tests/Client/Build.Desktop/QueryCountTests.cs]: >var query = context.Customer.IncludeTotalCount(true); [](start = 12, length = 53)

should have negative test case to test "false"

File: src/Microsoft.OData.Client/DataServiceQueryOfT.cs [hunk 5]
```diff
@@ -314,14 +314,27 @@ public DataServiceQuery<TElement> Expand<TTarget>(Expression<Func<TElement, TTar
         /// <returns>A new <see cref="T:Microsoft.OData.Client.DataServiceQuery`1" /> object that has the inline count option set.</returns>
         public DataServiceQuery<TElement> IncludeTotalCount()
```
Comment (xuzhg) [inline src/Microsoft.OData.Client/DataServiceQueryOfT.cs]: >IncludeTotalCount [](start = 42, length = 17)

TotalCount is OData V3 term, OData V4 is using Count. Consider to refactor as "IncludeCount()"?

File: src/Microsoft.OData.Client/DataServiceQueryOfT.cs (hunk 2 above)
Comment (KenitoInc) [reply src/Microsoft.OData.Client/DataServiceQueryOfT.cs]: Done

Comment (KenitoInc) [reply src/Microsoft.OData.Client/DataServiceQueryOfT.cs]: Switched to using `First`

File: src/Microsoft.OData.Client/DataServiceQueryOfT.cs (hunk 3 above)
Comment (KenitoInc) [reply src/Microsoft.OData.Client/DataServiceQueryOfT.cs]: Done

File: test/EndToEndTests/Tests/Client/Build.Desktop/QueryCountTests.cs (hunk 4 above)
Comment (KenitoInc) [reply test/EndToEndTests/Tests/Client/Build.Desktop/QueryCountTests.cs]: There is one

File: src/Microsoft.OData.Client/DataServiceQueryOfT.cs (hunk 5 above)
Comment (KenitoInc) [reply src/Microsoft.OData.Client/DataServiceQueryOfT.cs]: @xuzhg Do we still keep the IncludeTotalCount() for backwards compatibility? cc @mikepizzo

File: src/Microsoft.OData.Client/DataServiceQueryOfT.cs (hunk 1 above)
Comment (KenitoInc) [reply src/Microsoft.OData.Client/DataServiceQueryOfT.cs]: I have updated the code to ensure we add `$count=false` to the request Uri whenever we call `IncludeTotal(false)`

File: src/Microsoft.OData.Client/DataServiceQueryOfT.cs (hunk 2 above)
Comment (odero) [reply src/Microsoft.OData.Client/DataServiceQueryOfT.cs]: @KenitoInc That will just throw an exception if there's no "First". Might be better to just stick to `FirstOrDefault `and handle `mi == null`

File: src/Microsoft.OData.Client/DataServiceQueryOfT.cs (hunk 5 above)
Comment (mikepizzo) [reply src/Microsoft.OData.Client/DataServiceQueryOfT.cs]: Assuming that, for OData V4, setting IncludeTotalCount() sets $count=true, then we should keep it for backward compatibility.

Comment (xuzhg) [reply src/Microsoft.OData.Client/DataServiceQueryOfT.cs]: Or maybe we can mark the old one as [obsoleted] and add new one as "IncludeCount()". 

I agree both. 

---
In reply to: [381424145](https://github.com/OData/odata.net/pull/1658#discussion_r381424145) [](ancestors = 381424145)

File: test/EndToEndTests/Tests/Client/Build.Desktop/QueryCountTests.cs [hunk 6]
```diff
@@ -278,6 +293,28 @@ public void GetTotalCountInvalidTest()
...
+            {
+                var response = query.Execute() as QueryOperationResponse<Customer>;
+
+                var count = response.TotalCount;
+                Assert.Fail("Should fail on getting customer count");
+            }
+            catch (Exception ex)
+            {
+                Assert.AreEqual(true, ex.Message.Contains("Count value is not part of the response stream"));
+            }
```
Comment (xuzhg) [inline test/EndToEndTests/Tests/Client/Build.Desktop/QueryCountTests.cs]: We can't use "Assert.Throws(...)" assert?

File: src/Microsoft.OData.Client/DataServiceQueryOfT.cs (hunk 5 above)
Comment (KenitoInc) [reply src/Microsoft.OData.Client/DataServiceQueryOfT.cs]: @mikepizzo @xuzhg 
I have added `QueryOperationResponse<T>.Count` and `IncludeCount`
I have deprecated `QueryOperationResponse<T>.TotalCount` and `IncludeTotalCount`

File: test/FunctionalTests/Tests/DataOData/Tests/OData.Common.Tests/PublicApi/PublicApi.bsl [hunk 7]
```diff
@@ -7973,7 +7973,11 @@ public class Microsoft.OData.Client.MessageWriterSettingsArgs {
 }
 
 public class Microsoft.OData.Client.QueryOperationResponse : Microsoft.OData.Client.OperationResponse, IEnumerable {
+	long Count  { public virtual get; }
 	Microsoft.OData.Client.DataServiceRequest Query  { public get; }
+	[
+	ObsoleteAttribute(),
+	]
```
Comment (g2mula) [inline test/FunctionalTests/Tests/DataOData/Tests/OData.Common.Tests/PublicApi/PublicApi.bsl]: When adding the obsolete attribute, it is good to add a message to be displayed and whether to classify the "obsoletion" as an error or a warning ie
```[ObsoleteAttribute("This property is obsolete, please use Count", false)]```
and have the attribute in one line.
also please specify the access level of the prop: public, private, internal?

Comment (KenitoInc) [reply test/FunctionalTests/Tests/DataOData/Tests/OData.Common.Tests/PublicApi/PublicApi.bsl]: @g2mula This file is generated. It's used by the Astoria end to end tests

File: src/Microsoft.OData.Client/ALinq/ResourceBinder.cs [hunk 8]
```diff
@@ -1061,7 +1061,16 @@ private static Expression AnalyzeAddCountOption(MethodCallExpression mce, CountO
...
+
+            ConstantExpression countQueryOption = StripTo<ConstantExpression>(mce.Arguments[0]);
+            if ((bool)countQueryOption.Value == true)
+            {
+                rse.CountOption = CountOption.CountQueryTrue;
+            }
+            else
+            {
+                rse.CountOption = CountOption.CountQueryFalse;
+            }
```
Comment (habbes) [inline src/Microsoft.OData.Client/ALinq/ResourceBinder.cs]: What about when count option is neither true or false? i.e. when it should not be explicitly set in the query?

File: src/Microsoft.OData.Client/DataServiceQueryOfT.cs [hunk 9]
```diff
@@ -312,14 +312,41 @@ public DataServiceQuery<TElement> Expand<TTarget>(Expression<Func<TElement, TTar
...
         {
-            MethodInfo mi = typeof(DataServiceQuery<TElement>).GetMethod("IncludeTotalCount");
+            return this.IncludeCount(true);
+        }
+
+        /// <summary>Requests that the count of all entities in the entity set be returned inline with the query results.</summary>
+        /// <returns>A new <see cref="T:Microsoft.OData.Client.DataServiceQuery`1" /> object that has the inline count option set.</returns>
+        /// <param name="countQuery">Whether to include total count.</param>
+        [Obsolete("Please use includeCount(bool countQuery)")]
```
Comment (habbes) [inline src/Microsoft.OData.Client/DataServiceQueryOfT.cs]: use the right case of the method name in the Obsolete attribute message
```diff
- Please use includeCount(bool countQuery)
+ Please use IncludeCount(bool countQuery)
```

File: src/Microsoft.OData.Client/QueryOperationResponseOfT.cs [hunk 10]
```diff
@@ -42,6 +42,7 @@ internal QueryOperationResponse(HeaderCollection headers, DataServiceRequest que
 
         /// <summary>The server result set count value from a query, if the query has requested the value.</summary>
         /// <returns>The return value can be either zero or a positive value equal to the number of entities in the set on the server.</returns>
+        [Obsolete("Please use Count")]
         public override long TotalCount
         {
             get
```
Comment (habbes) [inline src/Microsoft.OData.Client/QueryOperationResponseOfT.cs]: I think you should change this getter to
```c#
[Obsolete(...)]
public override long TotalCount
{
  get
  {
    return Count;
  }
}
```

File: src/Microsoft.OData.Client/ALinq/ResourceBinder.cs (hunk 8 above)
Comment (KenitoInc) [reply src/Microsoft.OData.Client/ALinq/ResourceBinder.cs]: If it's not set in the query, the call doesn't reach here

File: src/Microsoft.OData.Client/DataServiceQueryOfT.cs (hunk 9 above)
Comment (KenitoInc) [reply src/Microsoft.OData.Client/DataServiceQueryOfT.cs]: fixed

File: src/Microsoft.OData.Client/QueryOperationResponseOfT.cs (hunk 10 above)
Comment (KenitoInc) [reply src/Microsoft.OData.Client/QueryOperationResponseOfT.cs]: fixed

File: src/Microsoft.OData.Client/ALinq/ResourceBinder.cs (hunk 8 above)
Comment (KenitoInc) [reply src/Microsoft.OData.Client/ALinq/ResourceBinder.cs]: `IncludeCount(false)` test case is included. #Resolved

**Pull request merged**